cmd/maya-apiserver/app/server/backup_endpoint_v1alpha1.go

```diff
@@ -0,0 +1,388 @@
+	bkp.Spec.PrevSnapName = lastsnap
+
+	glog.Infof("Creating backup %s for volume %q poolUUID:%v", bkp.Spec.SnapName,
+		bkp.Spec.VolumeName,
+		bkp.ObjectMeta.Labels["cstorpool.openebs.io/uid"])
+
+	_, err = openebsClient.OpenebsV1alpha1().BackupCStors(bkp.Namespace).Create(bkp)
+	if err != nil {
+		glog.Errorf("Failed to create backup: error '%s'", err.Error())
+		return nil, CodedError(500, err.Error())
+	}
+
+	glog.Infof("Backup resource:'%s' created successfully", bkp.Name)
+	return "", nil
+}
+
+// createSnapshotForBackup will create a snapshot for given backup
+func createSnapshotForBackup(bkp *v1alpha1.BackupCStor) error {
+	snapOps, err := snapshot.Snapshot(&v1alpha1.SnapshotOptions{
+		VolumeName: bkp.Spec.VolumeName,
+		Namespace:  bkp.Namespace,
+		CasType:    string(v1alpha1.CstorVolume),
+		Name:       bkp.Spec.SnapName,
+	})
+	if err != nil {
+		return CodedError(400, err.Error())
+	}
+
+	glog.Infof("Creating backup snapshot %s for volume %q", bkp.Spec.SnapName, bkp.Spec.VolumeName)
+
+	snap, err := snapOps.Create()
+	if err != nil {
+		glog.Errorf("Failed to create snapshot:%s error '%s'", bkp.Spec.SnapName, err.Error())
+		return CodedError(500, err.Error())
+	}
+	glog.Infof("Snapshot:'%s' created successfully for backup:%s", snap.Name, bkp.Name)
+	return nil
+}
+
+// loadClientFromServiceAccount loads a k8s and openebs client from a ServiceAccount
+// specified in the pod running
+func loadClientFromServiceAccount() (*internalclientset.Clientset, *kubernetes.Clientset, error) {
+	cfg, err := rest.InClusterConfig()
+	if err != nil {
+		glog.Errorf("Failed to fetch k8s cluster config. %+v", err)
+		return nil, nil, err
+	}
+
+	k8sClient, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		glog.Errorf("Failed to create k8s client: %v", err)
+		return nil, nil, err
+	}
+
+	openebsClient, err := internalclientset.NewForConfig(cfg)
+	if err != nil {
+		glog.Errorf("Failed to create openeEBS client. %+v", err)
+		return nil, nil, err
+	}
+
+	return openebsClient, k8sClient, nil
+}
+
+// findHealthyCVR will find a healthy CVR for a given volume
+func findHealthyCVR(openebsClient *internalclientset.Clientset, volume string) *v1alpha1.CStorVolumeReplica {
+	listOptions := v1.ListOptions{
+		LabelSelector: "openebs.io/persistent-volume=" + volume,
+	}
+
+	cvrList, err := openebsClient.OpenebsV1alpha1().CStorVolumeReplicas("").List(listOptions)
+	if err != nil {
+		return nil
+	}
+
+	// Select a healthy cvr for backup
+	for _, cvr := range cvrList.Items {
+		if cvr.Status.Phase == v1alpha1.CVRStatusOnline {
+			return &cvr
+		}
+	}
+
+	return nil
+}
+
+// getLastBackupSnap will fetch the last successful backup's snapshot name
+func getLastBackupSnap(openebsClient *internalclientset.Clientset, bkp *v1alpha1.BackupCStor) (string, error) {
+	lastbkpname := bkp.Spec.BackupName + "-" + bkp.Spec.VolumeName
+	b, err := openebsClient.OpenebsV1alpha1().BackupCStorLasts(bkp.Namespace).Get(lastbkpname, v1.GetOptions{})
+	if err != nil {
```

Comment:
you should do only for 'not found' error

Reply:
And, we are supposed to create/update this after the successful backup, not at the start of backup.

Reply:
ok